format/fit/fit.go

```diff
@@ -0,0 +1,337 @@
+			DecodeFn:    decodeFIT,
+		})
+
+	interp.RegisterFS(fitFS)
+}
+
+var fitCRCTable = [16]uint16{
+	0x0000, 0xCC01, 0xD801, 0x1400, 0xF001, 0x3C00, 0x2800, 0xE401,
+	0xA001, 0x6C00, 0x7800, 0xB401, 0x5000, 0x9C01, 0x8801, 0x4400,
+}
+
+func calcCRC(bytes []byte) uint16 {
+	var crc uint16
+	crc = 0
+	for i := 0; i < len(bytes); i++ {
+		// compute checksum of lower four bits of byte
+		var checkByte = bytes[i]
+		var tmp = fitCRCTable[crc&0xF]
+		crc = (crc >> 4) & 0x0FFF
+		crc = crc ^ tmp ^ fitCRCTable[checkByte&0xF]
+		tmp = fitCRCTable[crc&0xF]
+		crc = (crc >> 4) & 0x0FFF
+		crc = crc ^ tmp ^ fitCRCTable[(checkByte>>4)&0xF]
+	}
+
+	return crc
+}
+
+type fitContext struct {
+	dataSize   uint64
+	headerSize uint64
+}
+
+type dataRecordContext struct {
+	compressed         bool
+	data               bool
+	localMessageType   uint64
+	hasDeveloperFields bool
+}
+
+type fileDescriptionContext struct {
+	devIdx        uint64
+	fDefNo        uint64
+	typ           string
+	name          string
+	unit          string
+	nativeFieldNo uint64
+	nativeMsgNo   uint64
+}
+
+type devFieldDefMap map[uint64]map[uint64]mappers.FieldDef
+type localFieldDefMap map[uint64]map[uint64]mappers.FieldDef
+type localMsgIsDevDef map[uint64]bool
+
+func fitDecodeFileHeader(d *decode.D, fc *fitContext) {
+	frameStart := d.Pos()
+
+	headerSize := d.FieldU8("header_size")
+	d.FieldU8("protocol_version")
+	d.FieldU16("profile_version")
+	dataSize := d.FieldU32("data_size")
+
+	d.FieldRawLen("data_type", 4*8, d.AssertBitBuf([]byte(".FIT")))
+	if headerSize == 14 {
+		headerCRC := calcCRC(d.BytesRange(frameStart, int(headerSize)-2))
+		d.FieldU16("crc", d.UintValidate(uint64(headerCRC)))
+	}
+	fc.headerSize = headerSize
+	fc.dataSize = dataSize
+}
+
+func fitDecodeDataRecordHeader(d *decode.D, drc *dataRecordContext) {
+	headerType := d.FieldU1("header_type", scalar.UintMapSymStr{0: "normal", 1: "compressed"})
+	drc.compressed = headerType == 1
+	if drc.compressed {
+		localMessageType := d.FieldU2("local_message_type")
+		d.FieldU32("time_offset")
+		drc.localMessageType = localMessageType
+		drc.data = true
+	} else {
+		mTypeIsDef := d.FieldU1("message_type", scalar.UintMapSymStr{0: "data", 1: "definition"})
+		hasDeveloperFields := d.FieldBool("has_developer_fields")
+		d.FieldBool("reserved")
+		localMessageType := d.FieldU4("local_message_type")
+
+		drc.hasDeveloperFields = hasDeveloperFields
+		drc.localMessageType = localMessageType
+		drc.data = mTypeIsDef == 0
+	}
+}
+
+func fitDecodeDefinitionMessage(d *decode.D, drc *dataRecordContext, lmfd localFieldDefMap, dmfd devFieldDefMap, isDevMap localMsgIsDevDef) {
+	d.FieldU8("reserved")
+	endian := d.FieldU8("architecture")
+	switch endian {
+	case 0:
+		d.Endian = decode.LittleEndian
+	case 1:
+		d.Endian = decode.BigEndian
+	default:
+		d.Fatalf("Unknown endian %d", endian)
+	}
+	messageNo := d.FieldU16("global_message_number", mappers.TypeDefMap["mesg_num"])
+	if messageNo == 206 { // developer field_description
```

Comment:
I guess we could generate constants for these but i have a feeling very few of them would be used?

Reply:
I can make a manual constant for this "magic" number, and then have a gathered place for the if more are needed in the future.

Reply:
Sounds good